src/rebuild/ras.c

```diff
@@ -0,0 +1,60 @@
+/**
+ * (C) Copyright 2021 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * rebuild: RAS
+ *
+ * This file contains RAS helpers for rebuild events.
+ */
+#define D_LOGFAC       DD_FAC(rebuild)
+
+#include <daos_srv/daos_server.h>
+
+static int
+raise_ras(ras_event_t id, uuid_t *pool, char *op_str, char *msg)
+{
+	if ((pool == NULL) || uuid_is_null(*pool)) {
+		D_ERROR("invalid pool\n");
+		return -DER_INVAL;
+	}
+
+	ds_notify_ras_event(id, msg, RAS_TYPE_INFO, RAS_SEV_INFO,
+			    NULL /* hwid */, NULL /* rank */,
+			    NULL /* jobid */, pool, NULL /* cont */,
+			    NULL /* objid */, NULL /* ctlop */, op_str);
+	return 0;
+}
+
+int
+rebuild_notify_ras_start(uuid_t *pool, char *op_str)
+{
+	char *msg = "Pool rebuild started.";
+
+	return raise_ras(RAS_POOL_REBUILD_START, pool, op_str, msg);
+}
+
+int
+rebuild_notify_ras_end(uuid_t *pool, char *op_str, int op_rc)
+{
+	char	*msg, *status = NULL;
+	int	 rc = 0;
+
+	if (op_rc != 0) {
+		D_ASPRINTF(status, "failed: "DF_RC, DP_RC(op_rc));
+		if (status == NULL)
+			return -DER_NOMEM;
+	}
+	D_ASPRINTF(msg, "Pool rebuild %s", op_rc ? status : "finished.");
+	if (msg == NULL) {
+		rc = -DER_NOMEM;
+		goto out;
+	}
```

Comment:
use local variable like status[32], msg[32], to avoid these error check?

Reply:
I'd suggest msg be a bit longer (maybe 64?) but otherwise agree.

Reply:
I'll change it if you both feel strongly, but I'm curious as to what the downside of the dynamic allocation is. Using a fixed-length array seems brittle. If we change the message format in the future we'll have to be careful to make sure that the arrays are still sized correctly.